### PR TITLE
feat(bailian): add native support for Alibaba Cloud Bailian (百炼)

### PR DESCRIPTION
**Pull request by Barsit**
---
  ## Summary

  Adds **Alibaba Cloud Bailian (百炼)** (`dashscope.aliyuncs.com`) as a primary provider. Setting `BAILIAN_API_KEY` is
  the whole setup (`BAILIAN_BASE_URL` optional for region switching).

  ### Provider

  - New `internal/providers/bailian` package, type `bailian`, default base URL
  `https://dashscope.aliyuncs.com/compatible-mode/v1`, Bearer auth.
  - Chat completions, streaming, ListModels, Responses (via chat translation), Embeddings, Files, Batches, and
  Passthrough all delegate to `openai.CompatibleProvider`.
  - Embedding models (`text-embedding-v3`, `text-embedding-v4`) are served by the compatible-mode API but are not
  auto-discovered via `GET /v1/models` -- configure via `BAILIAN_MODELS`.

  ### `max_tokens` compatibility shim

  Bailian deprecated `max_tokens` on 2026-04-20 in favor of `max_completion_tokens`. The provider transparently maps
  `max_tokens` -> `max_completion_tokens` for every request. If the caller explicitly sets `max_completion_tokens`, that
  value is preserved and `max_tokens` is only stripped (not overridden).

  ### Multi-region

  Bailian is available in Beijing (default), Singapore, Frankfurt, and Hong Kong. `BAILIAN_BASE_URL` selects the region.

  ### Docs

  New `docs/providers/bailian.mdx` (`max_tokens` compatibility, multi-region base URLs, model IDs), wired into
  nav/overview; README table, `.env.template`, `config.example.yaml`, CLAUDE.md, Swagger `@description` updated.

  ## Testing

  - **31 unit tests** covering:
    - Happy paths: ChatCompletion, Streaming, ListModels, Embeddings, Responses, Passthrough, CreateBatch, GetBatch,
  ListBatches, CancelBatch, CreateFile, ListFiles, GetFile, DeleteFile, GetFileContent, GetBatchResults
    - Error paths: upstream 400/401/500 for ChatCompletion, StreamChatCompletion, Responses, Embeddings, ListModels,
  Passthrough; transport failure
    - Provider-specific: `max_tokens` -> `max_completion_tokens` mapping, explicit `max_completion_tokens` preservation,
  nil/empty edge cases, request cloning, field preservation, Registration type and default URL
  - **Coverage: 91.5%**
  - **E2E:** existing generic gateway tests pass; no provider-specific e2e fixtures pollute the shared test harness
  - **Manually verified** against the live Bailian API (`dashscope.aliyuncs.com`): chat completions (non-streaming +
  SSE), `/v1/responses`, and `/v1/embeddings` all confirmed working

  This PR replaces #390 and addresses the feedback from @SantiagoDePolonia (lint fix + merge main).

  🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Release Notes

* **New Features**
  * Added Alibaba Cloud Bailian (百炼) as a supported LLM provider (chat, streaming, embeddings, batches, and file handling), including passthrough routing.
  * Automatically converts Bailian `max_tokens` to `max_completion_tokens` for compatibility.

* **Documentation**
  * Added a Bailian provider guide with region-specific base URL examples and supported feature details.
  * Updated README/provider docs, configuration examples, and environment templates to include Bailian setup variables.  

* **Chores**
  * Updated ignore patterns for local pre-commit hook files.

### CODE REVIEW

**Comment by Barsit** [comment]
@SantiagoDePolonia This PR (#392) replaces #390.

Both review items from your feedback on #390 have been addressed:

1. ✅ **Fix the lint issue** — 2 unchecked `json.Unmarshal` errors in `bailian_test.go` are now handled with proper error checks.
2. ✅ **Merge main and fix conflicts** — Merged `upstream/main` (includes Xiaomi MiMo provider, cache logging fixes) and resolved all 5 file conflicts.

Thanks!

**Comment by Barsit** [comment]
@coderabbitai Thanks for the review.

Re: the 5 actionable comments:

1. **docs.go** — Swagger description missing Bailian. This is an auto-generated file (), not a source file we edit directly.
2. **redis.go** —  test. This is upstream code (cache fix from #391), not part of this PR's diff.
3. **chat_content.go** —  validation. This is upstream code (Xiaomi provider from #389), not part of this PR's diff.
4. **bailian_test.go** — Base URL assertion. ✅ Fixed: both  and  now verify  matches the expected constant.
5. **xiaomi/audio.go** — MIME type case sensitivity. This is upstream code (Xiaomi provider from #389), not part of this PR's diff.

**Comment by Barsit** [comment]
@coderabbitai Thanks for the follow-up review. All actionable items have been addressed:

- **`.env.template`**: Added `BAILIAN_MODELS` after `BAILIAN_BASE_URL`.
- **`config/config.example.yaml`**: Added `bailian` to `enabled_passthrough_providers`.
- **`bailian.go`**: `json.Marshal` and `MergeUnknownJSONFields` errors are now properly handled (returns original request on failure).
- **`bailian_test.go`**: Stream `io.ReadAll` errors are now checked with `t.Fatalf`; transport failure test uses a `RoundTripper` stub instead of port 1.
- Base URL assertions in `TestDefaultBaseURL` and `TestNew_UsesRegistrationAndDefaultBaseURL` were already fixed in the previous round.

File: internal/providers/bailian/bailian.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+// Embedding models (text-embedding-v3, text-embedding-v4) must be configured
+// via BAILIAN_MODELS as they are not auto-discovered from the upstream
+// /v1/models endpoint.
+func (p *Provider) Embeddings(ctx context.Context, req *core.EmbeddingRequest) (*core.EmbeddingResponse, error) {
+	return p.compatible.Embeddings(ctx, req)
+}
+
+// Passthrough routes an opaque provider-native request to Bailian.
+func (p *Provider) Passthrough(ctx context.Context, req *core.PassthroughRequest) (*core.PassthroughResponse, error) {
+	return p.compatible.Passthrough(ctx, req)
```
[the remark this answers is not in the document]
Comment (Barsit) [reply internal/providers/bailian/bailian.go:125]: 已修复 ✅ Passthrough 方法现在通过 `adaptPassthroughBody` 对 raw body 做 `max_tokens → max_completion_tokens` 转换，与 ChatCompletion/StreamChatCompletion 行为一致。新增了 3 个测试覆盖转换场景。

File: cmd/gomodel/docs/docs.go [hunk 2]
```diff
@@ -7009,7 +7009,7 @@ var SwaggerInfo = &swag.Spec{
 	BasePath:         "/",
 	Schemes:          []string{"http"},
 	Title:            "GoModel API",
-	Description:      "AI gateway routing requests to multiple LLM providers (OpenAI, Anthropic, Gemini, Groq, OpenRouter, DeepSeek, Z.ai, xAI, MiniMax, Oracle, Ollama). Drop-in OpenAI-compatible API.",
+	Description:      "AI gateway routing requests to multiple LLM providers (OpenAI, Anthropic, Gemini, Groq, OpenRouter, DeepSeek, Z.ai, xAI, MiniMax, Xiaomi MiMo, Oracle, Ollama). Drop-in OpenAI-compatible API.",
```
[the remark this answers is not in the document]
Comment (Barsit) [reply cmd/gomodel/docs/docs.go]: 已修复 ✅ Swagger 描述和 docs/openapi.json 的提供商列表中已添加 Bailian。

File: internal/providers/bailian/bailian_test.go [hunk 3]
```diff
@@ -0,0 +1,761 @@
...
+		t.Fatalf("status = %q, want completed", resp.Status)
+	}
+}
+
+func TestDefaultBaseURL(t *testing.T) {
+	provider := NewWithHTTPClient("key", nil, llmclient.Hooks{})
+	if provider == nil {
+		t.Fatal("expected non-nil provider")
+	}
+}
```
[the remark this answers is not in the document]
Comment (Barsit) [reply internal/providers/bailian/bailian_test.go:345]: 已修复 ✅ TestDefaultBaseURL 现在通过 mock server 发送实际请求并验证请求到达正确 host，不再仅检查非 nil。

File: internal/providers/bailian/bailian_test.go [hunk 4]
```diff
@@ -0,0 +1,771 @@
...
+	stream, err := provider.StreamChatCompletion(context.Background(), &core.ChatRequest{
+		Model:    "qwen3-flash",
+		Messages: []core.Message{{Role: "user", Content: "hi"}},
+	})
+	if err != nil {
+		t.Fatalf("StreamChatCompletion() error = %v", err)
+	}
+	defer stream.Close()
+
+	body, _ := io.ReadAll(stream)
```
[the remark this answers is not in the document]
Comment (Barsit) [reply internal/providers/bailian/bailian_test.go]: 已修复 ✅ 所有 `io.ReadAll` 和 `json.Marshal` 的错误现在都正确处理，不再用 `_` 忽略。

File: config/config.example.yaml [hunk 5]
```diff
@@ -199,6 +199,15 @@ providers:
...
 
+  bailian:
+    type: bailian
+    api_key: "${BAILIAN_API_KEY}"
+    # Base URL for Alibaba Cloud Bailian (DashScope).
+    # Default: Beijing - "https://dashscope.aliyuncs.com/compatible-mode/v1"
+    # Singapore: "https://{workspace-id}.ap-southeast-1.maas.aliyuncs.com/compatible-mode/v1"
+    # Frankfurt: "https://{workspace-id}.eu-central-1.maas.aliyuncs.com/compatible-mode/v1"
+    # Hong Kong: "https://{workspace-id}.cn-hongkong.maas.aliyuncs.com/compatible-mode/v1"
+
```
[the remark this answers is not in the document]
Comment (Barsit) [reply config/config.example.yaml:210]: 已修复 ✅ bailian 配置块现在包含 `# base_url: "https://dashscope.aliyuncs.com/compatible-mode/v1"` 注释行，与其他提供商风格一致。

File: internal/providers/bailian/bailian.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+		"max_completion_tokens": maxTokensJSON,
+	})
+	if err != nil {
+		return req
+	}
+	cloned := *req
+	cloned.ExtraFields = extra
+	cloned.MaxTokens = nil
+	return &cloned
+}
```
[the remark this answers is not in the document]
Comment (Barsit) [reply internal/providers/bailian/bailian.go:235]: 已修复 ✅ `adaptBailianRequest` 在 `json.Marshal` 和 `MergeUnknownJSONFields` 失败时现在通过 `slog.Warn` 记录警告，不再静默回退。

File: internal/providers/xiaomi/audio.go [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+	return nil, core.NewInvalidRequestError("file is required", nil)
+}
+
+// transcriptionMIMEType resolves the audio MIME type from the upload's content
+// type when present, falling back to the filename extension, then to wav.
+func transcriptionMIMEType(req *core.AudioTranscriptionRequest) string {
+	if contentType := strings.TrimSpace(req.FileContentType); strings.HasPrefix(contentType, "audio/") {
+		return contentType
+	}
+	switch strings.ToLower(path.Ext(strings.TrimSpace(req.Filename))) {
```
[the remark this answers is not in the document]
Comment (Barsit) [reply internal/providers/xiaomi/audio.go:206]: 感谢指出。MIME 参数剥离属于 Xiaomi provider 的范围，不在本次 Bailian PR 的范围内。会在后续 PR 中处理。

File: internal/core/chat_content.go [hunk 8]
```diff
@@ -27,10 +27,49 @@ type ImageURLContent struct {
...
+	if len(s) < len(scheme) || !strings.EqualFold(s[:len(scheme)], scheme) {
+		return false
+	}
+	meta, _, ok := strings.Cut(s[len(scheme):], ",")
+	if !ok {
+		return false
+	}
+	mediaType, _, _ := strings.Cut(meta, ";")
+	return strings.Contains(mediaType, "/")
+}
```
[the remark this answers is not in the document]
Comment (Barsit) [reply internal/core/chat_content.go:64]: 感谢指出。这是核心验证层的既有问题，不在本次 Bailian PR 的范围内。会在后续 PR 中处理。

[the remark this answers is not in the document]
Comment (Barsit) [reply internal/core/chat_content.go:64]: 感谢指出。这是核心验证层的既有问题，不在本次 Bailian PR 的范围内。会在后续 PR 中处理。

File: internal/providers/xiaomi/audio.go [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+		}
+	}
+	return nil, core.NewInvalidRequestError("file is required", nil)
+}
+
+// transcriptionMIMEType resolves the audio MIME type from the upload's content
+// type when present, falling back to the filename extension, then to wav.
+func transcriptionMIMEType(req *core.AudioTranscriptionRequest) string {
+	if contentType := strings.TrimSpace(req.FileContentType); strings.HasPrefix(contentType, "audio/") {
+		return contentType
```
[the remark this answers is not in the document]
Comment (Barsit) [reply internal/providers/xiaomi/audio.go:204]: 感谢指出。这是 Xiaomi provider 的问题，不在本次 Bailian PR 的范围内。会在后续 PR 中处理。

File: internal/cache/redis.go [hunk 10]
```diff
@@ -29,6 +29,7 @@ type RedisStore struct {
 
 // NewRedisStore creates a Redis-based key-value store.
 func NewRedisStore(cfg RedisStoreConfig) (*RedisStore, error) {
+	installRedisLogger()
```
[the remark this answers is not in the document]
Comment (Barsit) [reply internal/cache/redis.go:32]: 感谢指出。这是 cache/redis 层的问题，不在本次 Bailian PR 的范围内。会在后续 PR 中处理。

**Comment by SantiagoDePolonia** [comment]
Merging! I'll adjust it a little bit in the upcoming PR.
